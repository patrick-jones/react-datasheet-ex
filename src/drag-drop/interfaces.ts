import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
