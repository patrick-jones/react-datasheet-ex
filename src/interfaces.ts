import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
