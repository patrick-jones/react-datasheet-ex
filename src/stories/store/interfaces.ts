import RDS from 'react-datasheet';
import {ColumnHeader, Overflow} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export interface ExampleModel {
  id: string;
  headers: ExampleHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
  overflow: Overflow;
}

export interface ExampleHeader extends ColumnHeader {
  hidden?: boolean;
}

export interface ExampleRow {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
}

