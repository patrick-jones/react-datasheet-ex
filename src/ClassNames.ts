export const READ_ONLY_CELL = 'cell read-only';
export const ACTION_CELL = 'rdx-action-cell';

export const HEADER_CELL = 'rdx-header-cell';
export const HEADER_CELL_LEFT = 'rdx-header-cell--left';
export const HEADER_CELL_CENTER = 'rdx-header-cell--center';
export const HEADER_CELL_RIGHT = 'rdx-header-cell--right';

export const HEADER_CELL_CONTAINER = 'rdx-header-cell-container';
export const HEADER_CELL_TITLE = 'rdx-header-cell-title';

export const HEADER_DRAG_SOURCE = 'rdx-header-drag-source';
export const HEADER_DROP_TARGET = 'rdx-header-drop-target';

export const ROW = 'rdx-row';
export const ROW_DROP_TARGET = 'rdx-row-drop-target';
export const ROW_DRAG_SOURCE = 'rdx-row-drag-source';

export const RESIZING_WIDTH = 'rdx-ex-resizing-width';
export const DATA_GRID_RESIZABLE = 'rdx-data-grid-resizable';
export const WIDTH_RESIZE_HANDLE = 'rdx-width-resize-handle';

export const SCROLL_PANEL = 'rds-ex-scroll-panel';
