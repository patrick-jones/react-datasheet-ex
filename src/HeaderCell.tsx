import React, {PureComponent, ReactNode} from 'react';
import {ColumnHeader, HeaderRendererProps} from './interfaces';
import {READ_ONLY_CELL, HEADER_CELL} from './ClassNames';
import headerDefaults from './headerDefaults';


export default class HeaderCell<T extends ColumnHeader = ColumnHeader>
  extends PureComponent<HeaderRendererProps<T>> {

  render(): ReactNode {
    const {
      width,
      id,
      align,
      className,
      colSpan,
      overflow,
    } = headerDefaults(this.props.header, {overflow: this.props.overflow});
    const style = typeof width === 'undefined' ?
      this.props.style :
      Object.assign({}, (this.props.style || {}), {width: `${width}px`});

    const cn = [
      READ_ONLY_CELL,
      HEADER_CELL,
      `${HEADER_CELL}--${align}`,
      `${HEADER_CELL}--${overflow}`,
      className || '',
    ].filter(c => !!c).join(' ');

    return (
      <th
        key={id}
        style={style}
        colSpan={colSpan}
        className={cn}
      >
        {this.props.children}
      </th>
    );
  }
}
