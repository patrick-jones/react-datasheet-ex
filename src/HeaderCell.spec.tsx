import React from 'react';
import {mount, shallow} from 'enzyme';
import {ColumnHeader} from './interfaces';
import HeaderCell from './HeaderCell';
import {HEADER_CELL, HEADER_CELL_CENTER, HEADER_CELL_LEFT, HEADER_CELL_RIGHT} from './ClassNames';


describe('HeaderCell', () => {
  const header: ColumnHeader = {
    id: 'foo',
    title: 'AA',
  };

  it('Renders text as expected', () => {
    const wrapper = mount(
      <HeaderCell col={0} header={header}>{header.id}</HeaderCell>
    );
    expect(wrapper.find('th.rdx-header-cell').text()).toBe(header.id);
  });

  it('Renders style as needed', () => {
    const wrapper = shallow(
      <HeaderCell col={0} header={header}>{header.id}</HeaderCell>
    );
    expect(wrapper.find('th.rdx-header-cell').prop('style')).toBeUndefined();

    const styled = shallow(
      <HeaderCell col={0} header={{...header, width: 100}}>
        {header.id}
      </HeaderCell>
    );
    expect(styled.find('th.rdx-header-cell').prop('style'))
      .toEqual({width: '100px'});
  });

  it('Should render with alignment options', () => {
    const center = mount(
      <HeaderCell col={0} header={{...header, width: 100}}>
        {header.id}
      </HeaderCell>
    ).find(`th.${HEADER_CELL}.${HEADER_CELL_CENTER}`);

    expect(center.length).toBe(1);

    const left = mount(
      <HeaderCell col={0} header={{...header, width: 100, align: 'left'}}>
        {header.id}
      </HeaderCell>
    ).find(`th.${HEADER_CELL}.${HEADER_CELL_LEFT}`);

    expect(left.length).toBe(1);

    const right = mount(
      <HeaderCell col={0} header={{ id: 'h1', title: 'h1', align: 'right' }}>
        Content
      </HeaderCell>
    ).find(`th.${HEADER_CELL}.${HEADER_CELL_RIGHT}`);

    expect(left.length).toBe(1);
  });


  it('matches snapshot', () => {
    const wrapper = shallow(
      <HeaderCell col={0} header={header}>{header.id}</HeaderCell>
    );
    expect(wrapper).toMatchSnapshot();
  });

});
