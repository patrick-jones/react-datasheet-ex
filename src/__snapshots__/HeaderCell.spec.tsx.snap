// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderCell matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HeaderCell
    col={0}
    header={
      Object {
        "id": "foo",
        "title": "AA",
      }
    }
  >
    foo
  </HeaderCell>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "foo",
    "nodeType": "host",
    "props": Object {
      "children": "foo",
      "className": "cell read-only rdx-header-cell rdx-header-cell--center rdx-header-cell--clip",
      "colSpan": undefined,
      "style": undefined,
    },
    "ref": null,
    "rendered": "foo",
    "type": "th",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "foo",
      "nodeType": "host",
      "props": Object {
        "children": "foo",
        "className": "cell read-only rdx-header-cell rdx-header-cell--center rdx-header-cell--clip",
        "colSpan": undefined,
        "style": undefined,
      },
      "ref": null,
      "rendered": "foo",
      "type": "th",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
